### PR TITLE
Path to multigpu: Cuistl multigpu

### PR DESCRIPTION
**Pull request by kjetilly**
We are creating a new linear solver GPU path compatible with the DUNE ISTL library. This PR provides communication classes for the GPU for MPI, as well as a solver adapter to make the cuistl matrix and vector classes work with the FlexibleSolver-interface.

To run the code on GPUs, one simply needs to set the linear solver to `cubicgstab` and the preconditioner to `CUILU0`, in eg. a JSON file as:

```json
{
    "preconditioner" : {"type":"CUILU0"},
    "solver" : "cubicgstab"
}
```

[This is related to the issue 3932](https://github.com/OPM/opm-simulators/issues/3932)

[This PR partially replaces PR 4137](https://github.com/OPM/opm-simulators/pull/4137)

This PR builds upon [PR 4635](https://github.com/OPM/opm-simulators/pull/4635).

### CODE REVIEW

**Review by atgeirr — COMMENTED** [review]
Good!

File: opm/simulators/linalg/cuistl/SolverAdapter.hpp [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+    using typename Dune::IterativeSolver<X, X>::range_type;
+    using typename Dune::IterativeSolver<X, X>::field_type;
+    using typename Dune::IterativeSolver<X, X>::real_type;
+    using typename Dune::IterativeSolver<X, X>::scalar_real_type;
+    static constexpr auto block_size = domain_type::block_type::dimension;
+    using XGPU = Opm::cuistl::CuVector<real_type>;
+
+    // TODO: Use a std::forward
+    SolverAdapter(Operator& op,
+                  std::shared_ptr<Dune::ScalarProduct<X>> sp,
```
[the remark this answers is not in the document]
Comment (kjetilly) [reply opm/simulators/linalg/cuistl/SolverAdapter.hpp]: Changed reference for scalar product. For the preconditioner, I would prefer to change that to reference in a future commit, there was something about the ownership there I am not 100% sure of.

**Comment by atgeirr** [comment]
jenkins build this please

File: opm/simulators/linalg/PreconditionerFactory_impl.hpp [hunk 2]
```diff
@@ -224,6 +226,20 @@ struct StandardPreconditioners
             return std::make_shared<OwningTwoLevelPreconditioner<O, V, LevelTransferPolicy, Comm>>(op, prm, weightsCalculator, pressureIndex, comm);
           });
         }
+
+#if HAVE_CUDA
+        F::addCreator("CUILU0", [](const O& op, const P& prm, const std::function<V()>&, std::size_t, const C& comm) {
+            const double w = prm.get<double>("relaxation", 1.0);
+            static constexpr auto block_size = V::block_type::dimension;
```
[the remark this answers is not in the document]
Comment (kjetilly) [reply opm/simulators/linalg/PreconditionerFactory_impl.hpp]: Removed.

**Comment by kjetilly** [comment]
Jenkins build this please

**Comment by atgeirr** [comment]
All green, merging!

**Pull request merged**